src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs

```diff
@@ -160,46 +161,69 @@ public static object ConvertFromJson(string input, bool returnHashtable, int? ma
             error = null;
             try
             {
-                // JsonConvert.DeserializeObject does not throw an exception when an invalid Json array is passed.
-                // This issue is being tracked by https://github.com/JamesNK/Newtonsoft.Json/issues/1930.
-                // To work around this, we need to identify when input is a Json array, and then try to parse it via JArray.Parse().
 
-                // If input starts with '[' (ignoring white spaces).
-                if (Regex.Match(input, @"^\s*\[").Success)
-                {
-                    // JArray.Parse() will throw a JsonException if the array is invalid.
-                    // This will be caught by the catch block below, and then throw an
-                    // ArgumentException - this is done to have same behavior as the JavaScriptSerializer.
-                    JArray.Parse(input);
-
-                    // Please note that if the Json array is valid, we don't do anything,
-                    // we just continue the deserialization.
-                }
+                var serializer = new JsonSerializer();
 
-                var obj = JsonConvert.DeserializeObject(
-                    input,
-                    new JsonSerializerSettings
+                serializer.TypeNameHandling = TypeNameHandling.None;
+                serializer.MetadataPropertyHandling = MetadataPropertyHandling.Ignore;
+                serializer.MaxDepth = maxDepth;
+                using (System.IO.MemoryStream stream = new System.IO.MemoryStream(System.Text.Encoding.Default.GetBytes(input)))
+                {
+                    using (System.IO.StreamReader streamReader = new System.IO.StreamReader(stream))
                     {
-                        // This TypeNameHandling setting is required to be secure.
-                        TypeNameHandling = TypeNameHandling.None,
-                        MetadataPropertyHandling = MetadataPropertyHandling.Ignore,
-                        MaxDepth = maxDepth
-                    });
+                        using (JsonReader reader = new JsonTextReader(streamReader))
+                        {
+                            bool isArray = false;
+                            var readResult = reader.Read();
+                            // If the first token in our file is an array let's read in that token so that our next token is an object or a primitive.
+                            // This will allow newtonsoft to deserialize the incoming json one object at a time instead of doing the whole object at once.
+                            if (reader.TokenType == JsonToken.StartArray)
```

Comment:
Could this be something like (pseudocode):

```
if token type is startarray
    return ParseArray(serializer, reader, out error)
else
    return ParseStandard(serialize, reader, out error)

ParseStandard:
    call serializer.Deserialize
    switch obj
        each case returns a value, as in the original code

ParseArray
    elements = new List<object>()
    while reader.Read and token isn't end-array
        originalElement = serializer.Deserialize
        processedElement = ParseStandard(serializer, reader, out error)
        if error break
    return elements.ToArray
```

Reply:
My latest commit should refactor it to be more like this